core/domain/stats_services_test.py

```diff
@@ -45,532 +50,765 @@ def _get_continuous_computation_class(cls):
+    EXP_ID = 'exp_id0'
 
-        for ind in range(5):
-            event_services.StartExplorationEventHandler.record(
-                'eid', 1, exp.init_state_name, 'session_id_%s' % ind,
-                {}, feconf.PLAY_TYPE_NORMAL)
-            event_services.StateHitEventHandler.record(
-                'eid', 1, exp.init_state_name, 'session_id_%s' % ind,
-                {}, feconf.PLAY_TYPE_NORMAL)
-        event_services.AnswerSubmissionEventHandler.record(
-            'eid', 1, exp.init_state_name,
-            self.DEFAULT_RULESPEC_STR, self.DEFAULT_SESSION_ID,
-            self.DEFAULT_TIME_SPENT, self.DEFAULT_PARAMS, '1')
-        for _ in range(2):
-            event_services.AnswerSubmissionEventHandler.record(
-                'eid', 1, exp.init_state_name,
-                self.DEFAULT_RULESPEC_STR, self.DEFAULT_SESSION_ID,
-                self.DEFAULT_TIME_SPENT, self.DEFAULT_PARAMS, '2')
-        ModifiedStatisticsAggregator.start_computation()
-        self.process_and_flush_pending_tasks()
-        with self._get_swap_context():
-            self.assertEquals(
-                stats_services.get_state_improvements('eid', 1), [{
-                    'type': 'default',
-                    'rank': 3,
-                    'state_name': exp.init_state_name
-                }])
-
-    def test_single_default_rule_hit(self):
-        exp = exp_domain.Exploration.create_default_exploration('eid')
-        exp_services.save_new_exploration('fake@user.com', exp)
-        state_name = exp.init_state_name
+    def _get_submitted_answer_dict_list(self, state_answers):
+        submitted_answer_dict_list = (
+            state_answers.get_submitted_answer_dict_list())
+        for submitted_answer_dict in submitted_answer_dict_list:
+            del submitted_answer_dict['json_size']
+        return submitted_answer_dict_list
 
-        event_services.StartExplorationEventHandler.record(
-            'eid', 1, state_name, 'session_id', {}, feconf.PLAY_TYPE_NORMAL)
-        event_services.StateHitEventHandler.record(
-            'eid', 1, state_name, 'session_id', {},
-            feconf.PLAY_TYPE_NORMAL)
-        event_services.AnswerSubmissionEventHandler.record(
-            'eid', 1, state_name,
-            self.DEFAULT_RULESPEC_STR, 'session_id', self.DEFAULT_TIME_SPENT,
-            self.DEFAULT_PARAMS, '1')
-        ModifiedStatisticsAggregator.start_computation()
-        self.process_and_flush_pending_tasks()
+    def setUp(self):
+        super(RecordAnswerTests, self).setUp()
+        self.owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+        user_services.get_or_create_user(self.owner_id, self.OWNER_EMAIL)
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+        self.exploration = self.save_new_valid_exploration(
+            self.EXP_ID, self.owner_id, end_state_name='End')
+
+    def test_record_answer_without_retrieving_it_first(self):
+        # Do not assert the initial state since
```

Comment:
It's actually wrong; it might have been carried over from old tests or I was just mistaken when I wrote this. The old model had this problem (retrieval would add a new entity), but the new answer storage model does not.